lib/crawl.js

```diff
@@ -0,0 +1,135 @@
+/* eslint-disable promise/always-return */
+const { spawn } = require('node:child_process')
+const { mkdir, readFile, writeFile } = require('node:fs/promises')
+const { URL } = require('node:url')
+const { join } = require('node:path')
+const pretty = require('pretty-time')
+
+const outdir = 'public/archive'
+
+const catchAllFilesPath = [
+  '/sites/',
+  '/sfc/sites/'
+]
+
+const wgetArgs = [
+  '--adjust-extension',
+  '--compression', 'gzip',
+  '--convert-links',
+  '--directory-prefix', outdir,
+  '--page-requisites',
+  '--recursive',
+  '--timestamping',
+  '--user-agent', 'Archive-It',
+  '--include-directories', catchAllFilesPath.join(',')
+]
+
+module.exports = {
+  crawl,
+  catchAllFilesPath,
+  getRejected,
+  getSavedFiles
+}
+
+async function crawl (urlish, extraArgs = []) {
+  const url = new URL(urlish, 'https://sfgov.org')
+  const { hostname, pathname } = url
+  const logDir = join('logs', hostname, pathname)
+
+  await mkdir(logDir, { recursive: true })
+  await mkdir(outdir, { recursive: true })
+
+  const logPath = join(logDir, 'wget.log')
+  const rejectedLogPath = join(logDir, 'wget.rejected.tsv')
+
+  const args = [
+    ...wgetArgs,
+    '--output-file', logPath,
+    '--rejected-log', rejectedLogPath,
+    '--include-directories', pathname,
+    ...extraArgs,
+    url
+  ].map(String)
+
+  console.log('🦀 crawling: %s (log file: %s)', url, logPath)
+  // console.log('→ `wget %s`', args.map(shellQuote).join(' '))
+
+  return new Promise((resolve, reject) => {
+    const start = process.hrtime()
+    const proc = spawn('wget', args, {
```

Comment:
similar to overwhelming the remote server, if we spawn too many processes, we could overwhelm the host machine... and potentially for the larger sites as those take longer to process.

Reply:
Oh - this is kinda related to how many "crawls" we do from the the other side. So.. if we fire off 100 "crawls", and that means there'll be 100 procs, and if they're long living, could end up getting weird behaviors.

Yea, was thinking about the incremental crawls. I kinda assumed there would be no more content being added... but if that's not true, the timestamp stuff will be extra helpful for getting small diffs of content rather than redownloading the entire site every time.